Mathlib/CategoryTheory/Triangulated/Opposite/Functor.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2024 Sophie Morel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sophie Morel
+-/
+import Mathlib.CategoryTheory.Triangulated.Opposite.Pretriangulated
+
+/-!
+# Opposites of functors between pretriangulated categories,
+
+If `F : C ⥤ D` is a functor between pretriangulated categories, we prove that
+`F` is a triangulated functor if and only if `F.op` is a triangulated functor.
+In order to do this, we first show that a `CommShift` structure on `F` naturally
+gives one on `F.op` (for the shifts on `Cᵒᵖ` and `Dᵒᵖ` defined in
+`CategoryTheory.Triangulated.Opposite.Basic`), and we then prove
+that `F.mapTriangle.op` and `F.op.mapTriangle` correspond to each other via the
+equivalences `(Triangle C)ᵒᵖ ≌ Triangle Cᵒᵖ` and `(Triangle D)ᵒᵖ ≌ Triangle Dᵒᵖ`
+given by `CategoryTheory.Pretriangulated.triangleOpEquivalence`.
+
+-/
+
+namespace CategoryTheory
+
+variable {C D : Type*} [Category C] [Category D] [HasShift C ℤ] [HasShift D ℤ] (F : C ⥤ D)
+  [F.CommShift ℤ]
+
+namespace Pretriangulated.Opposite
+
+/-- If `F` commutes with shifts, so does `F.op`, for the shifts chosen on `Cᵒᵖ` in
+`CategoryTheory.Triangulated.Opposite.Basic`.
+-/
+noncomputable scoped instance commShiftOpInt : F.op.CommShift ℤ := by
+  letI F' : OppositeShift C ℤ ⥤ OppositeShift D ℤ := F.op
+  letI : F'.CommShift ℤ := F.commShiftOp ℤ
+  apply F'.commShiftPullback
+
+end Pretriangulated.Opposite
+
+namespace Functor
+
+open Category Limits Pretriangulated Opposite
+
+lemma oppositeCommShiftInt_iso_eq (n m : ℤ) (h : n + m = 0) :
+    Functor.commShiftIso (F := F.op) n = isoWhiskerRight (shiftFunctorOpIso C n m h) F.op ≪≫
+      NatIso.op (Functor.CommShift.iso (F := F) m).symm ≪≫
+        isoWhiskerLeft F.op (shiftFunctorOpIso D n m h).symm := by
+  ext
+  rw [@commShiftPullback_iso_eq Cᵒᵖ _ ℤ ℤ _ _  (AddMonoidHom.mk' (fun (n : ℤ) => -n)
+    (by intros; dsimp; omega)) (inferInstance : HasShift (OppositeShift C ℤ) ℤ)
+    (OppositeShift D ℤ) _ (inferInstance : HasShift (OppositeShift D ℤ) ℤ) F.op
+    (commShiftOp F ℤ) n m (eq_neg_of_add_eq_zero_right h)]
+  rfl
```

Comment:
Instead of this lemma, it would be better to add the following more basic lemmas (which will facilitate some proofs below):
```lean
I think you should add these lemmas :
```lean
lemma op_commShiftIso_hom_app (X : Cᵒᵖ) (n m : ℤ) (h : n + m = 0):
    (F.op.commShiftIso n).hom.app X =
      (F.map ((shiftFunctorOpIso C n m h).hom.app X).unop).op ≫
        ((F.commShiftIso m).inv.app X.unop).op ≫
        (shiftFunctorOpIso D n m h).inv.app (op (F.obj X.unop)) := by
  obtain rfl : m = -n := by omega
  rfl

-- similarly `op_commShiftIso_inv_app` (which can be deduced from the one above)

lemma shift_map_op {X Y : C} (f : X ⟶ Y) (n : ℤ) :
    (F.map f).op⟦n⟧' = (F.op.commShiftIso n).inv.app _ ≫
      (F.map (f.op⟦n⟧').unop).op ≫ (F.op.commShiftIso n).hom.app _ :=
  (NatIso.naturality_1 (F.op.commShiftIso n) f.op).symm

lemma map_shift_unop {X Y : Cᵒᵖ} (f : X ⟶ Y) (n : ℤ) :
    F.map ((f⟦n⟧').unop) = ((F.op.commShiftIso n).inv.app Y).unop ≫
      ((F.map f.unop).op⟦n⟧').unop ≫ ((F.op.commShiftIso n).hom.app X).unop := by
  simp [shift_map_op]
```

Reply:
Some proofs lost their last `rfl` step after I did this.
I was only able to use `shift_map_op` and `map_shift_unop` in one proof.
The proof of `triangleOpEquivalence_functor_naturality` still timeouts if I'm not careful.